block-layer/codecs/dag-cbor.md

```diff
@@ -4,38 +4,46 @@
 
 ## Format
 
-The CBOR IPLD format is called DagCBOR to disambiguate it from regular CBOR.
-Most CBOR objects are valid DagCBOR. The only hard restriction is that any field
-with the tag 42 must be a valid CID.
+The CBOR IPLD format is called DagCBOR to disambiguate it from regular CBOR. Most CBOR objects are valid DagCBOR. The primary differences are:
+  * tag `42` interpreted as CIDs
+  * maps may only be keyed by strings
+  * additional strictness requirements about valid data encoding forms
 
-## Link Format
+## Links
 
-As with all IPLD formats, DagCBOR must be able to encode [Links].
-In DagCBOR, links are [CIDs] encoded using the raw-binary identity [Multibase]. That Multibase prefix (`0x00`) *must not* be omitted. They are stored as byte-string type (major type 2), with the tag 42.
+As with all IPLD formats, DagCBOR must be able to encode [Links]. In DagCBOR, links are the binary form of a [CID] encoded using the raw-binary identity [Multibase]. That is, the Multibase identity prefix (`0x00`) is prepended to the binary form of a CID and this new byte array is encoded into CBOR as a byte-string (major type 2), with the tag `42`.
 
-(the inclusion of the Multibase exists for historical reasons)
+The inclusion of the Multibase prefix exists for historical reasons and the identity prefix *must not* be omitted.
 
-## Map Key Restriction
+## Map Keys
 
-In DagCBOR, map keys must be strings.
+In DagCBOR, map keys must be strings, as defined by the [IPLD Data Model].
 
-## Canonical DagCBOR
+## Strictness
 
-Canonical DagCBOR must:
+DagCBOR requires that there exist a single way of encoding any given object, and that encoded forms contain no superfluous data that may be ignored or lost in a round-trip decode/encode.
 
-1. Use no tags other than the CID tag (42). Other tags may be lost in
-   conversion.
-2. Use the [canonical CBOR](https://tools.ietf.org/html/rfc7049#section-3.9)
-   encoding.
+Therefore the DagCBOR codec must:
+
+1. Use no tags other than the CID tag (`42`). A valid DagCBOR encoder must not encode using any additional tags and a valid DagCBOR decoder must reject objects containing additional tags as invalid.
+2. Use the canonical CBOR encoding defined by the the suggestions in [section 3.9 of the CBOR specification]. A valid DagCBOR decoder should reject objects not following these restrictions as invalid. Specifically:
+   * Integer encoding must be as short as possible.
+   * The expression of lengths in major types 2 through 5 must be as short as possible.
+   * Floating point values must be encoded as the smallest of 16-, 32-, or 64-bit floating point that accurately represents the value, even for integral values.
+   * The keys in every map must be sorted lowest value to highest. Sorting is performed on the bytes of the representation of the keys.
+     - If two keys have different lengths, the shorter one sorts earlier;
```

Comment:
I'd just like to briefly note, somewhere, as an aside, that I'm not at all fond of this sorting rule.  It's the RFC, and it's what dag-cbor certainly does do in prod already, so yeah, we should spec this and this PR is good.

But I'm not fond of it.

https://play.golang.org/p/f0kkdXKppKF

The length rule makes very unnatural-looking outcomes.

Reply:
yeah, it's an odd one, I'd rather just rely on straight byte ordering regardless of length

Reply:
Agree with your comment in https://github.com/ipld/specs/pull/236#issuecomment-578592843 -- we should still specify this.  To no small extent, we're clarifying descriptions on stuff already shipped and in use that's generated data we expect to persist, and it definitely does this.

Just noting my laments.  And if we make another codec specification in the future, this sorting is something I would very much flag as worth reconsidering.

A useless comment, for *this* PR, perhaps, since our hands are tied.  But maybe also good contextual headspace to have, because it's another piece of kindling for the idea that not all codecs will agree on sorting orders.